pkg/cmd/admin/image/verify-signature.go

```diff
@@ -0,0 +1,366 @@
+	`)
+)
+
+type VerifyImageSignatureOptions struct {
+	InputImage        string
+	ExpectedIdentity  string
+	PublicKeyFilename string
+	PublicKey         []byte
+	Confirm           bool
+	Remove            bool
+	CurrentUser       string
+
+	Client client.Interface
+	Out    io.Writer
+	ErrOut io.Writer
+}
+
+func NewCmdVerifyImageSignature(name, fullName string, f *clientcmd.Factory, out, errOut io.Writer) *cobra.Command {
+	opts := &VerifyImageSignatureOptions{ErrOut: errOut, Out: out}
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s IMAGE [--confirm]", name),
+		Short:   "Verify that the given IMAGE signature is trusted",
+		Long:    verifyImageSignatureLongDesc,
+		Example: fmt.Sprintf(verifyImageSignatureExample, name),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(opts.Complete(f, cmd, args, out))
+			kcmdutil.CheckErr(opts.Run())
+		},
+	}
+
+	cmd.Flags().BoolVar(&opts.Confirm, "confirm", opts.Confirm, "If true, the result of the verification will be recorded to an image object.")
+	cmd.Flags().BoolVar(&opts.Remove, "remove", opts.Remove, "If set, all signature verifications will be removed from the given image.")
+	cmd.Flags().StringVar(&opts.PublicKeyFilename, "public-key", opts.PublicKeyFilename, "A path to a public GPG key to be used for verification.")
+	cmd.Flags().StringVar(&opts.ExpectedIdentity, "expected-identity", opts.ExpectedIdentity, "An expected image docker reference to verify.")
+	return cmd
+}
+
+func (o *VerifyImageSignatureOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, args []string, out io.Writer) error {
+	if len(args) != 1 {
+		return kcmdutil.UsageError(cmd, "exactly one image must be specified")
+	}
+	o.InputImage = args[0]
+	if len(o.ExpectedIdentity) == 0 {
+		return kcmdutil.UsageError(cmd, "the --expected-identity must be specified")
+	}
+	var err error
+
+	// If --public-key is provided only this key will be used for verification and the
+	// .gnupg/pubring.gpg will be ignored.
+	if len(o.PublicKeyFilename) > 0 {
+		if o.Remove {
+			return kcmdutil.UsageError(cmd, "cannot use public key when removing verification status")
+		}
+		if o.PublicKey, err = ioutil.ReadFile(o.PublicKeyFilename); err != nil {
+			return err
+		}
+	}
+	if o.Client, _, err = f.Clients(); err != nil {
+		return err
+	}
+	// Only make this API call when we are sure we will be writing validation.
+	if o.Confirm && !o.Remove {
+		if me, err := o.Client.Users().Get("~"); err != nil {
+			return err
+		} else {
+			o.CurrentUser = me.Name
+		}
+	}
+
+	return nil
+}
+
+// verifySignature verifies the image signature and returns the identity when the signature
+// is valid.
+func (o *VerifyImageSignatureOptions) verifySignature(pc *signature.PolicyContext, img *imageapi.Image, sigBlob []byte) (string, error) {
+	// Pretend that this is the only signature of img, and see what the policy says.
```

Comment:
> i thought we are not going to require policy.json here at all.

This is up to you.

The code as is right now includes https://github.com/openshift/origin/commit/ed7bc58edf668867e4bb95f977be1a0908e1fa00 , which uses `policy.json`.

https://github.com/openshift/origin/pull/13585#issuecomment-290427845 instead proposes another alternative, asking the user to provide a keyring and using only that one.

(The original PR was using the user’s default keyring by default; that’s _plausible_ but not  obviously the right thing to do: Habitual GPG users tend to have `keyserver-options auto-key-retrieve` in `~/.gnupg/gpg.conf`, i.e. the default keyring contains various keys with varying levels of trust for email and such, which tends to have very little intersection with trust for running containers. The current containers/image API does not make it trivial to use the signature verification code with “all of `~/.gnupg`” as the set of trusted keys, hence the two proposed alternatives above.)